### ISSUE TITLE
Improve ApplicableMethod

### ISSUE THREAD

**Issue by ssiccha** [gap-system/gap#2871]
This stems from the discussion in #2860.

- Suggestion by @fingolfin: add "options" (in the sense of https://www.gap-system.org/Manuals/doc/ref/chap8.html) to ApplicableMethod that one can use instead of those magic constants 1 to 6 in the third argument, in order to adjust the output.
  - maybe get rid of the printlevel parameter in the process?
- Get rid of the messages that concern the first applicable method in the output of `ApplicableMethod( String, [Group(())], 1, 2 );`. After all we were only asking for the second method.
- Some of the line breaks in `ApplicableMethod( String, [Group(())], 1, 2 );` are really awkward.
- Improve the `Skipped: ` messages e.g. in `ApplicableMethod( IsCyclic, [Group(())], 1, "all" );`. `Skipped` seems to imply "not applicable". Maybe instead print `Applicable method no. i.`.

 Doing `ApplicableMethod( String, [Group(())], 1, 2 );` yields:
```
#I  Searching Method for String with 1 arguments:
#I  Total: 70 entries
#I  Method 2: ``String'' at /home/sergio/projects/gap/lib/teaching.g:794 , value: 
10000
#I  Skipped:
#I  Method 
10
 : ``String: for a group with generators'' at /home/sergio/projects/gap/lib/grp.gi\
:4259 , value: 36
function( G ) ... end
```

Doing `ApplicableMethod( IsCyclic, [Group(())], 1, "all" );` yields:
```
#I  Searching Method for IsCyclic with 1 arguments:
#I  Total: 7 entries
#I  Method 1: ``IsCyclic: system getter'' at /home/sergio/projects/gap/lib/grp.gd:
157 , value: 2*SUM_FLAGS+38
#I  Skipped:
#I  Method 4: ``IsCyclic'' at /home/sergio/projects/gap/lib/grp.gi:35 , value: 36
#I  Skipped:
#I  Method 
5
 : ``IsCyclic: generic method for groups'' at /home/sergio/projects/gap/lib/grp.gi:
44 , value: 35
#I  Skipped:
#I  Method 
7
 : ``IsCyclic: default method requiring categories and checking properties'' at /ho\
me/sergio/projects/gap/lib/grp.gd:157 , value: 1
#I  Skipped:
[ function( obj ) ... end, function( G ) ... end, function( G ) ... end, 
  function( <1 unnamed arguments> ) ... end ]
```

**Comment by markuspf** [comment]
I am not a fan of options. I'd rather have an options record.

**Comment by hulpke** [comment]
`ApplicableMethod` was hacked together in an hour as a debugging tool. The printlevel is there to cater to cases of debugging method selection (or installation) itself (why does a method apply) but avoid swamping every call to it with pages of printout of known properties.

It would be nice to keep these features in some way, but interface and printing never got much love and surely would benefit from some care.

**Comment by cdwensley** [comment]
I find myself totally stymied.  The plan was to prepare a set of experimental functions for a new release of Utils, for example ApplicableMethodUtils; get comments and make revisions; then eventually copy these into methwhy.g (without the 'Utils' in the names).  But, with everything working in 4.9, on testing with the development version I find that METHODS_OPERATION has changed, writing 8 items per method rather than the 6 items in 4.9, so nothing works.  Give up for now?

**Comment by fingolfin** [comment]
`METHODS_OPERATION` is a low-level implementation detail of GAP, accessing it from user code is not recommended and likely to break in new GAP releases.

In GAP 4.10, we added `MethodsOperation` which still is undocumented, but due to its nature, can be used in non-library code safely, as it returns a list of records, which makes extensions painless.

If you want to stay compatible with GAP 4.9, you can use `BASE_SIZE_METHODS_OPER_ENTRY` to compute the size of entries in `METHOD_OPERATION(oper,n)` (note that this also depends on `n`!!!). Best to look at library code using `BASE_SIZE_METHODS_OPER_ENTRY` to see how to do that.

**Comment by cdwensley** [comment]
Yes, I discovered MethodsOperation earlier.  However, it's not just the size of the entries that has changed but (probably) the positions of the various fields.  Having thought some more (while walking my frustrations off!) I've probably learnt enough about the details in the past week to attempt a pull request directly on devel - some time next week.

**Comment by fingolfin** [comment]
If you use `MethodsOperation` than you don't need to worry about it.

**Comment by cdwensley** [comment]
PR #2948 is a first attempt at addressing this issue.
